### PR TITLE
feat: implement timer module

### LINKED ISSUE #104: Implement the prompt module for execution time
## Feature Request
Implement the prompt segment showing the execution time of the last command. By default, the module will only be shown if the execution time of the last command if it took more than 2 seconds to execute.

Much like the `jobs` module, each shell will likely have its own say to track and retrieve execution time, so information about the last command's execution may need to be passed in as a flag during `starship init`.

#### Describe the solution you'd like
The implementation used in spaceship is a good reference for this module:
https://github.com/denysdovhan/spaceship-prompt/blob/5f0a747171cd171c578f9359da517545f60cc1da/sections/exec_time.zsh

### PR DESCRIPTION
**Pull request by chipbuster**
This PR adds a module in starship that adds a job timer.

#### Description

Implement a timer module that takes a commandline argument, the number of seconds the last job took to complete, and displays it if appropriate.

Alters shell initialization files to compute this number using `date +%s` where needed.

Adds a config section to configure minimum amount of time before timer is shown (default is 2s)

#### Motivation and Context

Closes #104

#### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

#### 
Screenshots (if appropriate):

Top-to-bottom: bash, zsh, fish (zsh is slightly doctored because my old shell config kept throwing some lines in there)

image

#### How Has This Been Tested?

Tested by using in all three environments. Bash and Zsh proved surprisingly tricky to do correctly, so their shell configs are now fairly large.

Behavior is not 100% consistent between shells: when spamming empty commands, fish will continue to display the last command time, while bash and zsh will not.

Passes all integration tests.

- [x] I have tested using **MacOS**
- [x] I have tested using **Linux**
- [ ] I have tested using **Windows**

#### Checklist:

- [ ] I have updated the documentation accordingly.
- [x] I have updated the tests accordingly.
(But only in-module tests--don't know if there are tests elsewhere)

### CODE REVIEW

**Comment by chipbuster** [comment]
A couple of notes about this: first, I've utterly goobered the git stuff--it's showing 25 commits to merge in, some of which I'm pretty sure were already merged. If anyone knows how to fix this, please let me know.

Second, I'm not sure if the design of the rendering function is sane--I think it's nice, but it also has a `(&u64, &&str)` as the argument, which seems a little ugly.

Finally, there are two places where the things can silently fail:
 - If a negative or very large number is passed as a minimum for displaying time, it'll simply not display the time.
- If a bash shell already has `DEBUG` trapped, re-trapping it will break whatever was depending on it (and there's no way to un-trap that function). Since we can't do timing without the DEBUG trap, the timer is just silently disabled. This has already bitten me once in testing (I forgot I had a primitive timer running already), but I'm not sure if we want to display a warning on every launch.

File: src/init.rs [hunk 1]
```diff
@@ -59,3 +52,97 @@ pub fn init(shell_name: &str) {
...
+
+/* For zsh: preexec_functions and precmd_functions provide preexec/precmd in a
+   way that lets us avoid clobbering them.
+
+   Zsh quirk: preexec() is only fired if a command is actually run (unlike in
+   bash, where spamming empty commands still triggers DEBUG). This means a user
+   spamming ENTER at an empty command line will see increasing runtime (since
+   preexec never actually fires to reset the start time).
+
+   To fix this, only pass the time if STARSHIP_START_TIME is defined, and unset
```
Comment (matchai) [inline src/init.rs]: Nice 👍

File: src/modules/timer.rs [hunk 2]
```diff
@@ -0,0 +1,83 @@
+use crate::config::TableExt;
```
Comment (matchai) [inline src/modules/timer.rs]: In #116, I unify `crate::config::TableExt` and `crate::config::Config`.
Maybe we should try merging that in first to avoid creating conflicts.

File: src/modules/timer.rs [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+    let elapsed = arguments
+        .value_of("elapsed_time")
+        .unwrap_or("x") // not a time
+        .parse::<u64>()
+        .ok()?;
+
+    // Question: Is this a good series of calls to use? Is there a cleaner way to set types?
+    // Question: If the input config is negative, should I warn the user?
+    let config_min = context
+        .config
```
Comment (matchai) [inline src/modules/timer.rs]: The module already contains its config, so you can access it via `module.config`.
https://github.com/starship/starship/blob/master/src/context.rs#L102

File: src/modules/timer.rs [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+        .get_module_config("timer")
+        .and_then(|table| table.get_config("min_time"))
+        .and_then(|value| value.as_integer())
```
Comment (matchai) [inline src/modules/timer.rs]: Ideally, I think we should have a `get_as_int` accessor with the other config accessors:
https://github.com/starship/starship/blob/af1b66708abc462fe0db34083c3f6c1d790f61a3/src/config.rs#L12-L14 (this is on the branch where `TableExt` is unified with `Config`)

File: src/modules/timer.rs [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+        n => format!("{}{}", n, suffix),
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn test_10s() {
```
Comment (matchai) [inline src/modules/timer.rs]: 👍

File: src/modules/character.rs [hunk 6]
```diff
@@ -1,4 +1,5 @@
 use super::{Context, Module};
+use crate::config::{Config, TableExt};
```
Comment (matchai) [inline src/modules/character.rs]: Is this addition needed? 🤔

Comment (chipbuster) [reply src/modules/character.rs]: Whoa, how did that get there?

I might have fumbled a key combo in my editor, but that is definitely not needed.

File: src/init.rs [hunk 7]
```diff
@@ -59,3 +52,97 @@ pub fn init(shell_name: &str) {
...
+};
+precmd_functions+=(starship_precmd);
+preexec_functions+=(starship_preexec);
+STARSHIP_START_TIME="$(date +%s)";
+"##;
+
+/* Fish setup is simple because they give us CMD_DURATION. Hooray! */
+const FISH_INIT: &str = r##"
+function fish_prompt;
+    set -l elapsed (math --scale=0 "$CMD_DURATION / 1000");
```
Comment (matchai) [inline src/init.rs]: I would suggest adding:
```fish
set -l command_duration "$CMD_DURATION$cmd_duration"
```

Between fish 2.7 and 3.0, the casing for the variable was changed.

File: src/init.rs [hunk 8]
```diff
@@ -59,3 +52,97 @@ pub fn init(shell_name: &str) {
         print!("{}", script);
     }
 }
+
+/* For bash: we need to manually hook functions ourself: PROMPT_COMMAND will exec
+   right before the prompt is drawn, and any function trapped by DEBUG will exec
+   before a command is run. There is no way to avoid clobbering these functions,
+   so we will not hook them if they are already defined.
+
+   There is a preexec/precmd framework for bash out there: if we find the
```
[the remark this answers is not in the document]
Comment (chipbuster) [reply src/init.rs]: That's correct. Actually, even if `bash-preexec` is installed on its own, it won't do anything because apparently you're supposed to source it as the very last thing in your shell config. Presumably if you were *using* the functionality, you'd have those arrays defined.

Honestly I don't know what the best way to do this is. I really want to avoid clobbering any pre-existing functionality, but there's no way to do that in native bash while also installing our timing functions.

**Comment by matchai** [comment]
> I've utterly goobered the git stuff--it's showing 25 commits to merge in, some of which I'm pretty sure were already merged. If anyone knows how to fix this, please let me know.

This has happened to me more times than I can count.  😅
Don't worry about it. We'll be squash-merging your changes into a single commit.

File: src/modules/mod.rs [hunk 9]
```diff
@@ -28,6 +29,7 @@ pub fn handle<'a>(module: &str, context: &'a Context) -> Option<Module<'a>> {
         "git_status" => git_status::module(context),
         "username" => username::module(context),
         "battery" => battery::module(context),
+        "timer" => timer::module(context),
```
Comment (matchai) [inline src/modules/mod.rs]: Thoughts on naming this module `command_duration`?
I don't think timer is very clear.

**Comment by johnletey** [comment]
> it's showing 25 commits to merge in, some of which I'm pretty sure were already merged.

yes @chipbuster @matchai ... this happens when you're on a branch (not the master branch) that you have previously opened a PR for (that is now merged) ... and then you create a new branch off of this non-master branch

so your new branch has all of your old commits (which have now been merged into master), and all of your new commits

File: src/modules/timer.rs [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+
+    module.set_style(module_color);
+    module.new_segment("timer", &format!("took {}", render_time(elapsed)));
+    module.get_prefix().set_value("");
+
+    Some(module)
+}
+
+/// Render the time into a nice human-readable string
+fn render_time(raw_seconds: u64) -> String {
```
Comment (matchai) [inline src/modules/timer.rs]: We already have [`uom`](https://crates.io/crates/uom) as a sub-dependency of `battery`.
Maybe we should use that for zero-cost conversions here?

**Comment by chipbuster** [comment]
@johnletey I think I accidentally merged starship/master into my branch, instead of rebasing my branch onto it, and they diverged some time ago.

If we don't squash on merge, do you know how the process is handled? Will Github detect that some of the commits are duplicate and skip over them?

**Comment by matchai** [comment]
If we don't squash we'll get repeat commits. 🤷‍♂️

File: src/modules/mod.rs (hunk 9 above)
Comment (chipbuster) [reply src/modules/mod.rs]: Agreed on `timer`. Would `cmd_duration` also work?

Comment (matchai) [reply src/modules/mod.rs]: Yep, that should be fine.

File: src/modules/timer.rs (hunk 3 above)
Comment (chipbuster) [reply src/modules/timer.rs]: Hmm, I can't seem to access the config from within the module-generation function though, since it's private to Module.

```
field `config` of struct `module::Module` is private
```

Would adding an accessor to `Module` work? Or should I just make the config `pub`?

File: src/modules/timer.rs (hunk 4 above)
Comment (chipbuster) [reply src/modules/timer.rs]: Implemented. I've called it get_as_i64 since it wasn't clear to me initially what type of int would be returned.

File: src/modules/timer.rs (hunk 10 above)
Comment (chipbuster) [reply src/modules/timer.rs]: @matchai My impression after reading the `uom` doc is that it's great if we want to convert seconds into hours, or minutes into days, but if we have a number of seconds that we want to display as DD:HH:MM:SS, I don't see a way to do it aside from writing a custom `Formatter`, which would basically replicate what we're doing now, but in the library.

Is there something I missed while skimming the docs? I'm still not quite used to reading rust docs yet.

**Comment by johnletey** [comment]
> If we don't squash on merge, do you know how the process is handled? Will Github detect that some of the commits are duplicate and skip over them?

Oh boy ... not quite sure ... that's as far as my Github expertise goes!

**Comment by Zegnat** [comment]
> _[…]_ I've utterly goobered the git stuff--it's showing 25 commits to merge in, some of which I'm pretty sure were already merged. If anyone knows how to fix this, please let me know.

Assuming you have at least one branch that is up-to-date with starship/master, you can rebase against that one. I like to [setup an upstream remote](https://help.github.com/en/articles/configuring-a-remote-for-a-fork) for that, then you should be able to do something like:

```bash
git rebase -i upstream/master
```

With this git should already try and deduplicate commits, and you will be given the option to `drop` any commits you recognise as unnecessary.

File: src/modules/timer.rs (hunk 10 above)
Comment (matchai) [reply src/modules/timer.rs]: Looks like you're right. 👍

File: src/init.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (matchai) [reply src/init.rs]: Would it be worthwhile for us to include and source `bash-preexec` if it hasn't already been?

File: src/modules/cmd_duration.rs [hunk 11]
```diff
@@ -0,0 +1,83 @@
...
+pub fn module<'a>(context: &'a Context) -> Option<Module<'a>> {
+    let mut module = context.new_module("cmd_duration")?;
+    let arguments = &context.arguments;
+    let elapsed = arguments
+        .value_of("elapsed_time")
+        .unwrap_or("x") // not a time
+        .parse::<u64>()
+        .ok()?;
+
+    // Question: If the input config is negative, should I warn the user?
```
Comment (matchai) [inline src/modules/cmd_duration.rs]: This can be a TODO for when error handling is implemented.
For the time being, I'd add a debug log for it and return `None`.

File: src/init.rs [hunk 12]
```diff
@@ -59,3 +52,101 @@ pub fn init(shell_name: &str) {
...
+precmd_functions+=(starship_precmd);
+preexec_functions+=(starship_preexec);
+STARSHIP_START_TIME="$(date +%s)";
+"##;
+
+/* Fish setup is simple because they give us CMD_DURATION. Just account for name
+changes between 2.7/3.0 and do some math to convert ms->s and we can use it */
+const FISH_INIT: &str = r##"
+function fish_prompt;
+    let -l CMD_DURATION "$CMD_DURATION$cmd_duration"
```
Comment (matchai) [inline src/init.rs]:
```suggestion
    set -l CMD_DURATION "$CMD_DURATION$cmd_duration"
```

File: src/main.rs [hunk 13]
```diff
@@ -36,6 +36,13 @@ fn main() {
         )
         .required(true);
 
+    let cmd_duration_arg = Arg::with_name("elapsed_time")
```
Comment (matchai) [inline src/main.rs]: I think we should stick to `cmd_duration` here too.
As I understand it, this is a measure of the duration of the last command's execution time, not the time elapsed since the start of the last command; or am I misunderstanding how it is implemented?

Comment (chipbuster) [reply src/main.rs]: Yes, that's correct. I was just mentally doing a manual string replace and missed this part.

File: src/init.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (chipbuster) [reply src/init.rs]: Hmmm...now here's an interesting idea.

The big problem I see is that bash-preexec has to be [the absolute last thing sourced in the shell startup](https://github.com/rcaloras/bash-preexec/issues/97), or else it breaks. This would, in turn, require that we have `starship init` be the last line in the user's shell, and even then, `bash-preexec` will [*still* clobber any existing DEBUG traps](https://github.com/rcaloras/bash-preexec/issues/39).

Fundamentally, I don't think there's a way to avoid clobbering `DEBUG`/`PROMPT_COMMAND` in the bash startup because of the limitations of the shell. We can try to take the edge off of it, but there's simply no way to make it foolproof. In light of that, I think we have two options: we can make a best-effort attempt to compensate ourselves and (possibly noisily) disable modules if that fails, or we can try to provide a framework where the user can provide their own initializations.

[the remark this answers is not in the document]
Comment (chipbuster) [reply src/init.rs]: I expect this to be  a pretty rare occurrence--the two Bash config frameworks I'm able to find are [Oh-My-Bash](https://github.com/ohmybash/oh-my-bash) and [Bash-it](https://github.com/Bash-it/bash-it). Oh-my-bash doesn't hook `DEBUG` at all, while `Bash-it` only hooks it specifically to provide a `preexec`-like function (sadly, it uses different function names for that functionality than `bash-preexec`).

What I'm taking from above is that it's going to be pretty rare for a user to hook the DEBUG trap without being aware that they've done it, so if we warn them about it, they should usually know how to fix it. We could enhance this by adding something in our documentation if needed.

[the remark this answers is not in the document]
Comment (matchai) [reply src/init.rs]: At this point, I'm thinking we should probably remove Bash support for this feature and circle back to it once we have a robust solution figured out. The rest of the feature seems to work well. 🙂

[the remark this answers is not in the document]
Comment (chipbuster) [reply src/init.rs]: Seems like a plan.

I'll put in some code to have starship log a message when calling init bash and then strip out the init code.

Since we're going to be squash-committing, should I reserve the code somewhere in a comment? Or are we okay with losing the bash init code for now?

[the remark this answers is not in the document]
Comment (matchai) [reply src/init.rs]: Could you open an issue to add command duration support in Bash? We could keep the code for it there. 
I'm not at my computer right now, otherwise I'd take care of it.

File: src/modules/cmd_duration.rs [hunk 14]
```diff
@@ -0,0 +1,93 @@
...
+pub fn module<'a>(context: &'a Context) -> Option<Module<'a>> {
+    let mut module = context.new_module("cmd_duration")?;
+    let arguments = &context.arguments;
+    let elapsed = arguments
+        .value_of("cmd_duration")
+        .unwrap_or("invalid_time")
+        .parse::<u64>()
+        .ok()?;
+
+    let signed_config_min = context
```
Comment (matchai) [inline src/modules/cmd_duration.rs]: The `config_value_X` convenience functions should work here:

```rust
let signed_config_min = module
	.config_value_i64("min_time")
	.unwrap_or(2);
```

File: src/main.rs [hunk 15]
```diff
@@ -36,6 +36,13 @@ fn main() {
         )
         .required(true);
 
+    let cmd_duration_arg = Arg::with_name("cmd_duration")
+        .short("d")
+        .long("duration")
+        .value_name("DURATION")
+        .help("The amount of time since the start of the last command (s)")
```
Comment (matchai) [inline src/main.rs]:
```suggestion
        .help("The execution duration the last command, in seconds")
```

File: src/main.rs [hunk 16]
```diff
@@ -36,6 +36,13 @@ fn main() {
         )
         .required(true);
 
+    let cmd_duration_arg = Arg::with_name("cmd_duration")
+        .short("d")
+        .long("duration")
+        .value_name("DURATION")
```
Comment (matchai) [inline src/main.rs]:
```suggestion
        .value_name("CMD_DURATION")
```

File: src/main.rs [hunk 17]
```diff
@@ -36,6 +36,13 @@ fn main() {
         )
         .required(true);
 
+    let cmd_duration_arg = Arg::with_name("cmd_duration")
+        .short("d")
+        .long("duration")
```
Comment (matchai) [inline src/main.rs]:
```suggestion
        .long("cmd-duration")
```

File: Cargo.toml [hunk 18]
```diff
@@ -1,6 +1,6 @@
 [package]
 name = "starship"
-version = "0.4.0"
+version = "0.3.2"
```
Comment (matchai) [inline Cargo.toml]: Whoops! Looks like the version in `Cargo.toml` and `Cargo.lock` got changed.

File: src/modules/cmd_duration.rs (hunk 14 above)
Comment (chipbuster) [reply src/modules/cmd_duration.rs]: Oh yeah, that's much better.

Was this added recently or did I miss it when I was writing this?

File: Cargo.toml (hunk 18 above)
Comment (chipbuster) [reply Cargo.toml]: Ah, sorry. I had a bad feeling about that change while I was rebasing.

File: src/modules/cmd_duration.rs (hunk 14 above)
Comment (matchai) [reply src/modules/cmd_duration.rs]: Yeah, it was a recent addition in #120.

File: src/modules/cmd_duration.rs [hunk 19]
```diff
@@ -0,0 +1,89 @@
...
+pub fn module<'a>(context: &'a Context) -> Option<Module<'a>> {
+    let mut module = context.new_module("cmd_duration")?;
+    let arguments = &context.arguments;
+    let elapsed = arguments
+        .value_of("cmd_duration")
+        .unwrap_or("invalid_time")
+        .parse::<u64>()
+        .ok()?;
+
+    let signed_config_min = module.config_value_i64("min time").unwrap_or(2);
```
Comment (matchai) [inline src/modules/cmd_duration.rs]: No spaces allowed in TOML keys 😛 

```suggestion
    let signed_config_min = module.config_value_i64("min_time").unwrap_or(2);
```

**Comment by matchai** [comment]
I totally forgot that we also need to add some integration tests and documentation for this new module.
Would you like me to take care of it after this PR, or would you rather tackle it yourself?

The following would need doing:
- Integration tests would be added in a `cmd_duration.rs` file in `tests/testsuite`
- A new heading would be made in `docs/config/README.md`
- `README.md` would be updated to no longer say that this is a planned feature

**Comment by chipbuster** [comment]
I'd be happy to do it, if you're willing to deal with a couple more reviews. I'm also starting to think I should hold off on working on this until this conference is over (Friday) and I have a chance to make up some sleep--I'm made some awfully silly mistakes the last few days.

What would need to be added for integration tests? Would it just be a bunch of tests using `render_module` with args for various `cmd-duration`s?

**Comment by matchai** [comment]
Haha, no rush. I don't mind going through some more reviews. Enjoy your conference and we can take care of this later. 😄 

> What would need to be added for integration tests? Would it just be a bunch of tests using render_module with args for various cmd-durations?

Yep, that should be fine. This is what I was thinking:
- Testing the output at > and < the default cmd-duration criteria
- The output at > and < a cmd-duration criteria set by configuration

The recent `directory` tests are a good reference:
https://github.com/starship/starship/pull/120/files#diff-901574afaba53d9b8e5b1e81fe984069

File: docs/config/README.md [hunk 20]
```diff
@@ -366,3 +366,28 @@ The module will be shown if any of the following conditions are met:
...
 disabled = true
 ```
+
+## Command Duration
+
+The `cmd_duration` module shows how long the last command took to execute.
+The module will be shown only if the command took longer than two seconds, or
+the `min_time` config value, if it exists.
+
+NOTE: Command duration is currently **not supported in `bash`**. See
```
Comment (matchai) [inline docs/config/README.md]: As we're using Vuepress for our configuration, this may be a good place to put a "custom container: https://vuepress.vuejs.org/guide/markdown.html#custom-containers

I was thinking either a tip or a warning.

```
::: tip NOTE
Command duration is currently **not supported in `bash`**. See
[this issue](https://github.com/starship/starship/issues/124) for more details.
:::
```

File: docs/config/README.md [hunk 21]
```diff
@@ -366,3 +366,28 @@ The module will be shown if any of the following conditions are met:
...
+| ---------- | ------- | ----------------------------------- |
+| `disabled` | `false` | Disables the `cmd_duration` module. |
+| `min_time` | `2`     | Shortest duration to show time for. |
+
+### Example
+
+```toml
+# ~/.config/starship.toml
+
+[username]
```
Comment (matchai) [inline docs/config/README.md]: Whoops!

```suggestion
[cmd_duration]
```

File: docs/config/README.md [hunk 22]
```diff
@@ -366,3 +366,28 @@ The module will be shown if any of the following conditions are met:
...
+the `min_time` config value, if it exists.
+
+NOTE: Command duration is currently **not supported in `bash`**. See
+[this issue](https://github.com/starship/starship/issues/124) for more details.
+
+### Options
+
+| Variable   | Default | Description                         |
+| ---------- | ------- | ----------------------------------- |
+| `disabled` | `false` | Disables the `cmd_duration` module. |
```
Comment (matchai) [inline docs/config/README.md]: Total nitpick: can we move `disabled` to the bottom of the table so it is consistent with the other modules?

File: src/modules/cmd_duration.rs [hunk 23]
```diff
@@ -0,0 +1,93 @@
...
+
+use super::{Context, Module};
+
+/// Outputs the time it took the last command to execute
+///
+/// Will only print if last command took more than a certain amount of time to
+/// execute. Default is two seconds, but can be set by config option `min_time`.
+pub fn module<'a>(context: &'a Context) -> Option<Module<'a>> {
+    let mut module = context.new_module("cmd_duration")?;
+    if module.config_value_bool("disabled").unwrap_or(false) {
```
Comment (matchai) [inline src/modules/cmd_duration.rs]: This logic is already present in `context.new_module()` for all modules. 😜
https://github.com/starship/starship/blob/f40f73dc8e26e6d8730344f4c6cbb47ee3d48342/src/context.rs#L104-L109

File: tests/testsuite/cmd_duration.rs [hunk 24]
```diff
@@ -0,0 +1,83 @@
...
+        .output()?;
+    let actual = String::from_utf8(output.stdout).unwrap();
+
+    let expected = format!("{} ", Color::Yellow.bold().paint("took 5s"));
+    assert_eq!(expected, actual);
+    Ok(())
+}
+
+#[test]
+#[ignore]
```
Comment (matchai) [inline tests/testsuite/cmd_duration.rs]: Because the config files that are generated are tempfiles created in a temp directory, there's no harm with running these tests on all machines. 🙂

```suggestion
```

File: tests/testsuite/cmd_duration.rs [hunk 25]
```diff
@@ -0,0 +1,83 @@
...
+        .output()?;
+    let actual = String::from_utf8(output.stdout).unwrap();
+
+    let expected = "";
+    assert_eq!(expected, actual);
+    Ok(())
+}
+
+#[test]
+#[ignore]
```
Comment (matchai) [inline tests/testsuite/cmd_duration.rs]:
```suggestion
```

File: tests/testsuite/cmd_duration.rs [hunk 26]
```diff
@@ -0,0 +1,83 @@
...
+        .output()?;
+    let actual = String::from_utf8(output.stdout).unwrap();
+
+    let expected = format!("{} ", Color::Yellow.bold().paint("took 10s"));
+    assert_eq!(expected, actual);
+    Ok(())
+}
+
+#[test]
+#[ignore]
```
Comment (matchai) [inline tests/testsuite/cmd_duration.rs]:
```suggestion
```

File: tests/testsuite/cmd_duration.rs [hunk 27]
```diff
@@ -0,0 +1,83 @@
+use ansi_term::Color;
```
Comment (matchai) [inline tests/testsuite/cmd_duration.rs]: The tests look great! 👍

File: docs/config/README.md (hunk 20 above)
Comment (chipbuster) [reply docs/config/README.md]: Great idea. I had initially bolded that text only because I wanted it to be attention-grabbing and bolding the whole thing seemed obnoxious, but a warning works much better for that.

File: src/modules/cmd_duration.rs (hunk 23 above)
Comment (chipbuster) [reply src/modules/cmd_duration.rs]: d'oh

File: tests/testsuite/cmd_duration.rs (hunk 24 above)
Comment (chipbuster) [reply tests/testsuite/cmd_duration.rs]: So the `#[ignore]` is usually used for tests that might (significantly) alter the test computer's state?

Comment (matchai) [reply tests/testsuite/cmd_duration.rs]: Yep. Ignored tests are run in Docker with `./integration_test`, but are not run with `cargo test`.

File: docs/config/README.md [hunk 28]
```diff
@@ -366,3 +366,30 @@ The module will be shown if any of the following conditions are met:
 [username]
 disabled = true
 ```
+
+## Command Duration
```
Comment (matchai) [inline docs/config/README.md]: Nitpick: The configuration option headings are listed in alphabetical order

File: src/init.rs [hunk 29]
```diff
@@ -58,3 +51,62 @@ pub fn init(shell_name: &str) {
...
+STARSHIP_START_TIME="$(date +%s)";
+"##;
+
+/* Fish setup is simple because they give us CMD_DURATION. Just account for name
+changes between 2.7/3.0 and do some math to convert ms->s and we can use it */
+const FISH_INIT: &str = r##"
+function fish_prompt;
+    set -l CMD_DURATION "$CMD_DURATION$cmd_duration";
+    set -l starship_duration (math --scale=0 "$CMD_DURATION / 1000");
+    starship prompt --status=$status --cmd-duration=$starship_duration;
```
Comment (matchai) [inline src/init.rs]: Unfortunately, it looks like a bug snuck in here. 🐛 
Because we are running the `set` commands before the prompt, we lose the old `$status` to the one set by `set`.

It seems we may need to set a variable at the start of the init script to hold onto the `$status` at the time:
```fish
set -l exit_code $status
```

Comment (chipbuster) [reply src/init.rs]: Funnily enough, I fixed that for bash and zsh. I'm not sure whether I missed the fact that it was occurring for fish or just forgot to make the edits.

Is there a way we could do automated tests in an actual shell to catch these sort of regressions? I was kind of doing them manually, but even with the basic combinations of `true`/`false` and `sleep` and multiple shells, it was getting a little difficult to figure out if the behavior I was seeing was correct.

Comment (matchai) [reply src/init.rs]: Agreed. We should introduce full end-to-end tests to catch these bugs. 👍

**Comment by matchai** [comment]
@all-contributors please add @chipbuster for docs and test.
